skimage/metrics/set_metrics.py

```diff
@@ -56,27 +129,72 @@ def hausdorff_distance(image0, image1, method="standard"):
     b_points = np.transpose(np.nonzero(image1))
 
-    # Handle empty sets properly:
-    # - if both sets are empty, return zero
-    # - if only one set is empty, return infinity
-    if len(a_points) == 0:
-        return 0 if len(b_points) == 0 else np.inf
-    elif len(b_points) == 0:
+    return _hausdorff_distance_set(a_points, b_points, method)
+
+
+def hausdorff_distance_mask(image0, image1, method='standard'):
+    """Calculate the Hausdorff distance between the contours of two segmentation masks.
+
+    Parameters
+    ----------
+    image0, image1 : ndarray
+        Arrays where ``True`` represents a pixel from a segmented object. Both arrays must have the same shape.
+    method : {'standard', 'modified'}, optional, default = 'standard'
+        The method to use for calculating the Hausdorff distance.
+        ``standard`` is the standard Hausdorff distance, while ``modified``
+        is the modified Hausdorff distance.
+    Returns
+    -------
+    distance : float
+        The Hausdorff distance between coordinates of the segmentation mask contours in
+        ``image0`` and ``image1``, using the Euclidean distance.
+    Notes
+    -----
+    The Hausdorff distance [1]_ is the maximum distance between any point on the
+    contour of ``image0`` and its nearest point on the contour of ``image1``, and
+    vice-versa.
+    The Modified Hausdorff Distance (MHD) has been shown to perform better
+    than the directed Hausdorff Distance (HD) in the following work by
+    Dubuisson et al. [2]_. The function calculates forward and backward
+    mean distances and returns the largest of the two.
+    References
+    ----------
+    .. [1] http://en.wikipedia.org/wiki/Hausdorff_distance
+    .. [2] M. P. Dubuisson and A. K. Jain. A Modified Hausdorff distance for object
+       matching. In ICPR94, pages A:566-568, Jerusalem, Israel, 1994.
+       :DOI:`10.1109/ICPR.1994.576361`
+       http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.1.8155
+    Examples
+    --------
+    >>> ground_truth = np.zeros((10, 10), dtype=bool)
+    >>> predicted = ground_truth.copy()
+    >>> ground_truth[2:9, 2:9] = True
+    >>> predicted[4:7, 2:9] = True
+    >>> hausdorff_distance_mask(ground_truth, predicted)
+    2.0
+    """
+
+    if method not in ('standard', 'modified'):
+        raise ValueError(f'unrecognized method {method}')
+    if image0.shape != image1.shape:
+        raise ValueError(f'shape of image0 {image0.shape} and image1 {image1.shape} should be equal.')
+
+    # Handle empty images
+    if not np.any(image0):  # no nonzero elements in image0
+        return 0.0 if not np.any(image1) else np.inf
+    elif not np.any(image1):
         return np.inf
 
-    fwd, bwd = (
-        cKDTree(a_points).query(b_points, k=1)[0],
-        cKDTree(b_points).query(a_points, k=1)[0],
-    )
+    a_points = np.concatenate(find_contours(image0>0))
+    b_points = np.concatenate(find_contours(image1>0))
```

Comment:
Currently, it seems like `hausdorff_distance_mask` and `hausdorff_distance` are a bit to similar for my taste. `hausdorff_distance_mask` just seems to add a call to `find_contours`. I'd rather just educate users in how to calculate the contours themselves. Additionally, if both functions arrive at a set of coordinates at some point it would make more sense to me to somehow update our API to accept coordinates directly as input. We probably have to add a second function here because it's difficult to update the existing `hausdorff_distance` accordingly.

Reply:
The main difference is that, in `hausdorff_distance`, we are limited to integer pixel positions (since we need to use the "contour image"), and the contour coordinates are just the indices of the nonzero pixels, whereas with the proposed `hausdorff_distance_mask` we can have the subpixel coordinates from `find_contours`.

I think the difficult part is that, IMO, the second option is what most users would intuitively expect, but changing the behaviour of the current function wouldn't be a great idea and would break stuff for those who used it correctly.

To go towards your suggestion, I guess we could use the `_hausdorff_distance_set` method and make it public (maybe as `hausdorff_distance_from_coordinates` or something like that), with the code from `hausdorff_distance_mask` put into the gallery and/or the docstring to show how to use it from the segmentation mask with `find_contours`.

So that would mean, to summarize:

- Remove `hausdorff_distance_mask` method
- Rename `_hausdorff_distance_set` and make it public (e.g. as `hausdorff_distance_from_coordinates`
- Put example code using `find_contours` in this `hausdorff_distance_from_coordinates` docstring and update the gallery example accordingly.

Would that work better?